Make "long type" printing type aware and trim types in E0275

estebank opened the pull request:
Instead of simple string cutting, use a custom printer to hide parts of long printed types.

On E0275, check for type length before printing.

estebank committed:
Make "long type" printing type aware

Instead of simple string cutting, use a custom printer to hide parts of
long printed types.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3475647053/jobs/5810097612) [(plain)](https://github.com/rust-lang/rust/commit/2696a3945fb1400ea358788a54dd49e88c5f285b/checks/9514521731/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_monomorphize/src/collector.rs at line 555:
     if s.chars().nth(33).is_some() {
         let shrunk = format!("{}", ty::ShortInstance(instance, 4));
         if shrunk == s {
-            return (s, None)
+            return (s, None);
 
 
         let path = tcx.output_filenames(()).temp_path_ext("long-type.txt", None);
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_monomorphize/src/collector.rs" "/checkout/compiler/rustc_index/src/vec/tests.rs" "/checkout/compiler/rustc_symbol_mangling/src/errors.rs" "/checkout/compiler/rustc_symbol_mangling/src/test.rs" "/checkout/compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs" "/checkout/compiler/rustc_symbol_mangling/src/legacy.rs" "/checkout/compiler/rustc_symbol_mangling/src/typeid.rs" "/checkout/compiler/rustc_index/src/interval.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

estebank committed:
On overflow errors, do not print out long types

estebank committed:
On E0275 do not print out the full type in the msg

When printing requirement overflow errors, do not print out the full
type name when it is longer than 50 characters long.

estebank committed:
Only use `...` instead of `_` for type elision

`_` might confuse people into believing that the type isn't known,
while `...` is not used anywhere else for types and is not valid
syntax, making it more likely to convey the right understanding.

oli-obk commented on compiler/rustc_middle/src/ty/error.rs:
```diff
@@ -985,6 +990,38 @@ fn foo(&self) -> Self::T { String::new() }
         false
     }
 
+    pub fn short_ty_string(self, ty: Ty<'tcx>) -> Result<String, (String, PathBuf)> {
```
Considering both callers to this function convert the result into an `(String, Option<PathBuf>)` maybe make that the return type of this function?

oli-obk reviewed (changes requested):
r=me with nit resolved

oli-obk commented:
r? @oli-obk

estebank committed:
review comment

estebank commented:
@bors r=oli-obk

bors commented:
:pushpin: Commit ec27227139812ddd22dc5294149047b42fde22ef has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk ec27227139812ddd22dc5294149047b42fde22ef -->
<!-- homu: {"type":"Approved","sha":"ec27227139812ddd22dc5294149047b42fde22ef","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

estebank commented:
@bors r=oli-obk

bors commented:
:pushpin: Commit bcb2655a9a3af867c3e48a332870acc36d92df88 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk bcb2655a9a3af867c3e48a332870acc36d92df88 -->
<!-- homu: {"type":"Approved","sha":"bcb2655a9a3af867c3e48a332870acc36d92df88","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
